rfcs/2022-06-13-db-sync.md

```diff
@@ -0,0 +1,62 @@
+# DB access and processing messages for iOS notification service extension
+
+## Problem
+
+The only way to receive/process notificaitons is via a separate NSE process that requires a concurrent DB and network access.
+
+SQLite concurenncy does not work, so we need to sync database access.
+
+The problem is complex, as we do not directly control db access from the app, it can be triggered by the message arriving and it may fail to complete in case the app is suspended in the background.
+
+So we need to prevent db access from starting when we know the app is about to be suspended.
+
+The last problem is how to receive and process messages in NSE - should it use recently added GET command or should it subscribe to the connections that receive messages and process messages normally.
+
+To summarize, 2 problems need to be solved:
+
+1. sync db access between 2 (or more, if we add share extension) processes
+
+2. prevent access from starting when the process is due to suspend, only complete operations.
+
+3. Receiveing and processing agent messages in NSE
+
+## Proposed solution
+
+For problem 1, we can use Posix semaphores from our core code, in the same bracket that manages database connection - it would wait for semaphore to be free and unlock it once the db operation is complete.
+
+For problem 2, we need to communicate from the app when it goes to the background to prevent database access starting and completing before the suspension. This would set some `aboutToSuspend` STM flag (or the opposite) that would prevent operations from progressing (using STM retry, that would block until the flag has the value allowing operation to progress).
+
+Several possibilities can be considered:
+
+- use this flag in the bracket that provides DB connections. While simple, it may lock some operations in the middle and may also lead to the situation when network operation succeeds but database access was blocked, and the database is not updated.
+- use this flag to stop network operations that would require database updates - like sending messages, subscriptions and ACK - all these operations would require database access once they succeed.
+- use two flags, for both cases above, but set them at different times since going to background - block new network operations as soon as the app goes to the background and block database access once the app is about to be suspended.
+
+The last option seems more robust. To do it, there will be an api allowing the app to communicat its phase:
+
+- app going to the background would trigger blocking new network operations and start a new background task - `background` phase.
+- background task receiving completion warning, or, maybe, some time after it is started - probably 20 seconds - or whatever happens earlier - would trigger call blocking db access - `suspended` phase.
```

Comment:
should existing background tasks also send `/_app phase suspended` in case they receive it earlier?